feat: Функциональность «Отзывы»

Serminos opened the pull request.

Commit:
feat: Функциональность «Отзывы»

naviwe commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,87 @@
+    @GetMapping
+    public List<ReviewDto> getReviews(@RequestParam(defaultValue = "0") Long filmId,
+                                      @RequestParam(defaultValue = "10") Long count) {
+        if (filmId == null) {
```
Проверка на null лишняя, т.к. filmId = 0 по дефолту

griatva commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,87 @@
+@Validated
+public class ReviewController {
+    private static final Logger log = LoggerFactory.getLogger(ReviewController.class);
+    ReviewService reviewService;
```
Мне кажется, стоит проставлять модификаторы доступа, например, в данном случае "private final"

griatva commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,87 @@
+
+    @PostMapping
+    public ReviewDto create(@Valid @RequestBody ReviewDto reviewDto) {
+        log.debug("Создание отзывы [{}]", reviewDto);
```
мелочь, конечно, но не "отзывы", а "отзыва"

griatva commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,87 @@
+
+    @PutMapping
+    public ReviewDto update(@Valid @RequestBody ReviewDto reviewDto) {
+        log.debug("Обновление отзывы [{}]", reviewDto);
```
мелочь, конечно, но не "отзывы", а "отзыва"

griatva commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,87 @@
+    public ReviewDto getReviewById(@PathVariable Long id) {
+        log.debug("Получение отзыва [{}]", id);
+        ReviewDto reviewDtoCreated = reviewService.getReviewById(id);
+        return reviewDtoCreated;
```
тут можно одной строчкой, без объявления переменной 
return reviewService.getReviewById(id);

griatva commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,87 @@
+    @GetMapping
+    public List<ReviewDto> getReviews(@RequestParam(defaultValue = "0") Long filmId,
+                                      @RequestParam(defaultValue = "10") Long count) {
+        if (filmId == null) {
```
да, похоже, что либо тут должна быть проверка на 0, либо по дефолту должен быть null.

Serminos replied:
Если не указывать (defaultValue = "0"), то по умолчание параметр считается обязательным
required = false . Так будет правильней указать и после проверять на null.
Исправлено

griatva replied:
да, но у тебя по тз в этом методе - если не указан id фильма, тогда вывести все отзывы. Если у тебя не указан в запросе id фильма, тогда он у тебя по умолчанию станет 0. Значит условие filmId == null - всегда будет false. Разве нет?

Serminos replied:
Да это так и потому required = false и не указываю defaultvalue. Сейчас ещё не закоммитил...

griatva commented on src/main/java/ru/yandex/practicum/filmorate/dto/ReviewDto.java:
```diff
@@ -0,0 +1,26 @@
+@AllArgsConstructor
+public class ReviewDto {
+    Long reviewId;
+    @NotBlank(message = "Отзыв не может быть пустым")
```
мне, кстати, кажется, что сам отзыв (его содержание) может быть пустым, а вот boolean isPositive должен быть not null.

Serminos replied:
Отзыв без текста, думаю все же текст должен быть. Я бы даже ввел ограничение чтобы минимальная длинна какая-то была. А то заспамить можно после. Проверку для Типа отзыва добавил

griatva replied:
Ну про наличие текста в отзыве - вопрос интерпретации. Не принципиально, наверное.

griatva commented on src/main/java/ru/yandex/practicum/filmorate/dto/ReviewDto.java:
```diff
@@ -0,0 +1,26 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class ReviewDto {
+    Long reviewId;
```
мне кажется, у всех полей должны быть модификаторы доступа private

griatva commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,87 @@
+    @DeleteMapping("/{id}")
+    public void delete(@PathVariable Long id) {
+        log.debug("Удаление отзыва [{}]", id);
+        reviewService.deleteById(id);
```
мне кажется, что было бы намного читабельнее, если бы методы назывались одинаково - либо delete, либо deleteById

Serminos replied:
Согласен. Delete лучше писать когда передается сущность для удаления. Переименовал в deleteById

griatva commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    public void deleteById(long reviewId) {
+        reviewStorage.deleteReview(reviewId);
```
то же самое - уже третье название "одного и того же" метода. мне кажется, что их стоит назвать одинаково - deleteById, иначе сложно ориентироваться

griatva commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,87 @@
+    @PutMapping("/{id}/like/{userId}")
+    public void addReviewLike(@PathVariable long id, @PathVariable long userId) {
+        log.debug("Добавление лайка к ревью [{}] - пользователем [{}]", id, userId);
+        reviewService.addLike(id, userId);
```
с названиями методов у меня везде одно и то же замечание - мне кажется важным, чтобы они назывались одинаково во всех слоях.

Serminos replied:
Это спорное утверждение. Каждый слой отражает свои сущности. И название методов может не совпадать. И методы могут использоваться разными контроллерами. Название должно отражать действие. 
На уровне контроллера действие в том чтобы добавить к ревью лайк. Потому и функция так называется.
Однако на уровне сервиса reviewService дописывать reviewService.addReviewLike похоже на тавтологию. 2 раза фигурирует Review

griatva commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,155 @@
+        return reviewDtos;
+    }
+
+    public List<ReviewDto> all(long limit) {
```
кажется, этот метод имеет много общего с методом public List<ReviewDto> getAllReview(long count), чуть выше в этом же классе, может, их объединить?

Serminos replied:
оставлен один метод

griatva commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,155 @@
+        return reviewDtos;
+    }
+
+    public void clear() {
```
не понимаю, где это используется?

griatva commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/h2/ReviewDbStorage.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public List<Review> findByFilmId(Long filmId, long count) {
+        return jdbcTemplate.query("SELECT * FROM review WHERE film_id = ?",
```
возможно, я неправильно поняла ТЗ, но, мне показалось, что списки отзывов должны быть отсортированы по их рейтингу (который useful). В тз написано вот так: "Отзывы должны сортироваться по рейтингу полезности." Звучит неоднозначно и больше нигде вроде не используется, ни в одном из ендпоинтов.. можно спросить наставника, что надо по итогу

Serminos replied:
да оставим этот вопрос на завтра. Не указан порядок сортировки и есть интересный момент к обсуждению связанный с этим

Serminos replied:
исправил

griatva commented on src/main/java/ru/yandex/practicum/filmorate/storage/impl/h2/ReviewLikeDbStorage.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Override
+    public ReviewLike update(ReviewLike reviewLike) {
```
в этом классе как будто несколько методов, которые не используются? (мне без IDEA сложно увидеть точно)
методы update, clear - они используются?

Serminos replied:
на уровне сервиса есть метод очистки - clear. update - сделан на всякий случай

griatva reviewed (changes requested):
круто сделано! 
я отметила, что смогла заметить, возможно, где-то непринципиальные моменты или ты совсем не согласен - тогда можем оставить на суд главного ревьюера)

Commit:
fix: исправления после review

- рефакторинг наименований
- рефакторинг модификаторов доступов
- убраны дублирующийся функции

Commit:
fix: useful ordering
